src/NuGet.Clients/NuGet.VisualStudio.Implementation/Telemetry/ExtensibilityTelemetryCollector.cs

```diff
@@ -0,0 +1,249 @@
+
+                // IVsSolutionRestoreService4
+                [nameof(IVsSolutionRestoreService4) + "." + nameof(IVsSolutionRestoreService4.RegisterRestoreInfoSourceAsync)] = new Count(),
+
+                // IVsNuGetProjectUpdateEvents
+                [nameof(IVsNuGetProjectUpdateEvents) + "." + nameof(IVsNuGetProjectUpdateEvents.SolutionRestoreStarted)] = new Count(),
+                [nameof(IVsNuGetProjectUpdateEvents) + "." + nameof(IVsNuGetProjectUpdateEvents.SolutionRestoreFinished)] = new Count(),
+                [nameof(IVsNuGetProjectUpdateEvents) + "." + nameof(IVsNuGetProjectUpdateEvents.ProjectUpdateStarted)] = new Count(),
+                [nameof(IVsNuGetProjectUpdateEvents) + "." + nameof(IVsNuGetProjectUpdateEvents.ProjectUpdateFinished)] = new Count(),
+            };
+        }
+
+        public TelemetryEvent ToTelemetryEvent()
+        {
+            TelemetryEvent data = new("extensibility");
+
+            foreach ((string api, Count count) in _counts)
+            {
+                if (count.Value > 0)
+                {
+                    data[api] = count.Value;
+                }
+            }
+
+            return data;
+        }
+
+        public void Dispose()
+        {
+            _eventListener?.Dispose();
+            _eventListener = null;
+
+            GC.SuppressFinalize(this);
+        }
+
+        private class Count
+        {
+            public int Value;
+        }
+
+        private class ExtensibilityEventListener : EventListener
+        {
+            private ExtensibilityTelemetryCollector _collector;
+            private Guid _expectedEtwSourceGuid;
+
+            public ExtensibilityEventListener(ExtensibilityTelemetryCollector collector)
+            {
+                _collector = collector;
+            }
+
+            protected override void OnEventSourceCreated(EventSource eventSource)
+            {
+                if (eventSource.Name == NuGetETW.ExtensibilityEventSourceName)
+                {
+                    _expectedEtwSourceGuid = eventSource.Guid;
+                    EnableEvents(eventSource, EventLevel.LogAlways);
+                }
+            }
+
+            protected override void OnEventWritten(EventWrittenEventArgs eventData)
+            {
+                if (eventData.EventSource.Guid != _expectedEtwSourceGuid)
+                {
+                    // My understanding of https://docs.microsoft.com/en-us/dotnet/api/system.diagnostics.tracing.eventlistener.oneventwritten?view=netframework-4.7.2
+                    // is that this should not be possible, yet when PrefView is collecting a trace, we get here.
+                    return;
+                }
+
+                var opcode = eventData.Opcode;
+                if (opcode == EventOpcode.Start || opcode == NuGetETW.CustomOpcodes.Add || opcode == EventOpcode.Info)
+                {
+                    if (_collector._counts.TryGetValue(eventData.EventName, out Count count))
```

Comment:
Is it worth trying to make this an atomic operation?  You could call `AddOrUpdate()` with a bogus add value that tells you it wasn't already in the dictionary so you could detect events that aren't blessed.  I'm not sure what would be best approach.

```c#
// Call returns -1 if something was added to our pre-populated dictionary which isn't allowed
if (_collector._counts.AddOrUpdate(eventData.EventName, -1, (key, value) => value + 1) == -1)
{
    if (eventData.EventName != "EventSourceMessage")
    {
        Debug.Assert(false, "VS Extensibility API without counter");
    }
    else
    {
        Debug.Assert(false, eventData.Message);
    }
}
```

Reply:
I think the comment at the top of the file (lines 18 to 29) mostly address this. If you don't agree with something in the comment, then we can discuss those points. Basically, `ConcurrentDictionary` makes runtime unnecessarily slower due to needing to use locks, whereas `Interlocked.Increment` is lock-free but also thread-safe.

Reply:
Yeah I had read the comment and was mostly thinking about atomic operations.  But in this implementation, you're not really using the dictionary for locking since its pre-populated so its not really doing two operations so you're right, you won't need anything but `Interlocked.Increment()`